uobtheatre/bookings/forms.py

```diff
@@ -0,0 +1,140 @@
+        return Ticket(
+            seat_group=SeatGroup.objects.get(id=self.seat_group_id),
+            concession_type=ConcessionType.objects.get(id=self.concession_type_id),
+            seat=Seat.objects.get(id=self.seat_id)
+            if self.seat_id is not None
+            else None,
+        )
+
+
+class TicketInputField(Field):
+    pass
+
+
+@convert_form_field.register(TicketInputField)
+def convert_form_field_to_string(field):
+    return graphene.List(
+        TicketInputType, description=field.help_text, required=field.required
+    )
+
+
+class BookingForm(MutationForm):
+    """Form for creating/updating a booking"""
+
+    tickets = TicketInputField(required=False)
+    user_email = CharField(required=False)
+
+    def clean(self):
+        """Validate form data on clean"""
+        cleaned_data = super().clean()
+        ## Various fields
+        if not self.instance.creator_id:
+            self.instance.creator = self.user
+
+        if cleaned_data.get("user_email"):
+            target_user, _ = User.objects.get_or_create(
+                email=cleaned_data.get("user_email"),
+                defaults={"first_name": "Anonymous", "last_name": "User"},
+            )
+        else:
+            target_user = self.user
+
+        if cleaned_data.get("user_email") or not self.instance.user_id:
+            self.instance.user = target_user
+
+        ## Check tickets
+        if cleaned_data.get("tickets") is not None:
+            ticket_objects = list(
+                map(lambda ticket: ticket.to_ticket(), cleaned_data.get("tickets"))
+            )
+
+            (
+                add_tickets,
+                delete_tickets,
+                total_number_of_tickets,
+            ) = self.instance.get_ticket_diff(ticket_objects)
+
+            self.cleaned_data["add_tickets"] = add_tickets
+            self.cleaned_data["delete_tickets"] = delete_tickets
+
+            if (
+                total_number_of_tickets > max_tickets_per_booking()
+                and not self.user.has_perm(
+                    "boxoffice", cleaned_data.get("performance").production
+                )
+            ):
+                raise ValidationError(
+                    {
+                        "tickets": "You may only book a maximum of %s tickets"
+                        % max_tickets_per_booking()
+                    }
+                )
+
+            # Check the capacity of the show and its seat_groups
+            err = cleaned_data.get("performance").check_capacity(ticket_objects)
+            if err:
+                raise ValidationError({"tickets": err})
+
+    def save(self, *args, **kwargs):
+        """Overrides the default form save to setup users (from provided emails)"""
+        if not self.instance.id or self.cleaned_data["user_email"]:
+            # Delete existing draft bookings
+            bookings = self.instance.user.bookings
+            if self.instance.id:
```

Comment:
Can you use something like?
```suggestion
            if not self.is_creation:
```

Reply:
Again, this is on a form. `is_creation` isn't a thing here but I guess I could implement it on a new custom base form class.